### DISCUSSION TITLE
basicManipulationIpc.insertGeometricElement led to the closure of the DB connection.

### DISCUSSION THREAD

**Discussion by VWlf** [iTwin/itwinjs-core#6756] [Q&A]
Currently, I am using @editor-frontend version 3.7.5 and @editor-backend version 4.0.3. When I finish creating a new element, the DB connection linked to the model always gets closed. Could this issue be due to the backend upgrade? If the frontend were upgraded, would it potentially resolve this error? Additionally, I noticed on the official website that insertGeometricElement has been removed after version 4.5.0. Are there any alternative methods available in version 4.0.3? I would greatly appreciate any guidance on these matters.
![111111111111](https://github.com/iTwin/itwinjs-core/assets/49948146/3343b5ac-b194-4cc5-83bf-03662382566c)
![2222222222](https://github.com/iTwin/itwinjs-core/assets/49948146/73dca9bc-d58b-496b-ae84-cd4957d6ff28)
![33333333333333333](https://github.com/iTwin/itwinjs-core/assets/49948146/c1fca7ba-f2eb-436a-b7a5-bbe2d220249e)

**Comment by pmconne** [comment]
> I am using @editor-frontend version 3.7.5 and @editor-backend version 4.0.3

You cannot mix different major versions.

**Reply by VWlf** [reply]
When I attempt to upgrade the front-end code to version 4.0.3 to align it with the back-end, this error persists. Are there any official, up-to-date examples available for reference? I'd like to compare and see if I'm making a mistake in my API calls.
![11111111111111](https://github.com/iTwin/itwinjs-core/assets/49948146/f1257d5d-ab87-4921-b0ea-4496678d55cf)

**Reply by VWlf** [reply]
I would like to know what value the iModelKey parameter holds in this method. Does it represent the absolute path for file storage, or is it any parameter that can access the model database connection, such as a VersionGuid?
![222222222222](https://github.com/iTwin/itwinjs-core/assets/49948146/f23d505a-f42d-49c4-92b5-40d5d4b7d79b)

**Reply by bbastings** [reply]
The optional "data" arg was removed because it's part of GeometricElementProps:
 
```
 /** Create and insert a new geometric element.
   * @param props Properties for the new [GeometricElement]($backend)
   * @see [GeometryStream]($docs/learning/common/geometrystream.md), [ElementGeometry]($common)
   * @throws [[IModelError]] if unable to insert the element
   */
  insertGeometricElement(props: GeometricElementProps): Promise<Id64String>;

export interface GeometricElementProps extends ElementProps {
  /** The id of the category for this geometric element. */
  category: Id64String;
  /** The geometry stream properties */
  geom?: GeometryStreamProps;
  /** How to build the element's GeometryStream. This is used for insert and update only. It is not a persistent property. It will be undefined in the properties returned by functions that read a persistent element. It may be specified as an alternative to `geom` when inserting or updating an element.
   * @beta
   */
  elementGeometryBuilderParams?: ElementGeometryBuilderParams;
  /** The placement properties */
  placement?: PlacementProps;
}
```

If you are using CreateElementWithDynamicsTool, doCreateElement was also changed to remove the data arg as it will be specified as the "elementGeometryBuilderParams" member of the supplied GeometricElementProps.

You should also be including the override keyword for methods you override to make sure that you are still overriding a method on the base class.

Anyway, if you are indeed using CreateElementWithDynamicsTool just do this:

```
  protected override async doCreateElement(props: GeometricElementProps): Promise<void> {
    try {
      this._startedCmd = await this.startCommand();
      if (undefined === props.id)
        await basicManipulationIpc.insertGeometricElement(props);
      else
        await basicManipulationIpc.updateGeometricElement(props);
      await this.saveChanges();
    } catch (err) {
      IModelApp.notifications.outputMessage(new NotifyMessageDetails(OutputMessagePriority.Error, BentleyError.getErrorMessage(err) || "An unknown error occurred."));
    }
  }
```

If you are not using CreateElementWithDynamicsTool and elementGeometryBuilderParams has not already been supplied via the GeometricElementProps, just do this:

`props.elementGeometryBuilderParams = { entryArray: data };`

HTH

**Comment by bbastings** [comment]
> I noticed on the official website that insertGeometricElement has been removed after version 4.5.0

It has not been removed, the change notice is in regards to the removal of a redundant parameter that should be supplied via ElementProps.

**Comment by bbastings** [comment]
> I would like to know what value the iModelKey parameter holds in this method.

It's the key associated with the call to open the iModel...

```
/**
 * A key used to identify an opened [IModelDb]($backend) between the frontend and backend for Rpc and Ipc communications.
 * Keys must be unique - that is there can never be two IModelDbs opened with the same key at any given time.
 * If no key is supplied in a call to open an IModelDb, one is generated and returned.
 * It is only necessary to supply a key if you have some reason to assign a specific value to identify an IModelDb.
 * If you don't supply the key, you must use the returned value for Rpc and Ipc communications.
 * @public
 */
export interface OpenDbKey {
  readonly key?: string;
}
```

**Reply by VWlf** [reply]
Thank you for your response. Over the past few days, through troubleshooting, I discovered that after saving the segmentation results, it is necessary to refresh the cache by closing the original link. The issue lies within the backend dependency, `@/editor-backend`, where the `openFile` method uses the `key` as the `filepath` parameter. Currently, the `key` does not represent an absolute path, which leads to the file not being found, preventing the link from being reopened. This in turn causes the `key` to become invalid and results in subsequent 404 errors from the API.

**Reply by kabentley** [reply]
I realize that you closed this issue, but I don't understand what you're seeing and your analysis of the problem doesn't make sense to me. Are you using `BriefcaseConnection.openFile`? 

It does not use the fileName as the key, and I don't see how you can possibly get the result you describe above. Can you show how you open the iModel?

**Reply by VWlf** [reply]
Apologies, it seems I misspoke. I'm referring to the code segment that enables the opening of an editable IModel.
`BriefcaseConnection.openStandalone`

**Reply by kabentley** [reply]
That doesn't use the filename for the key either. The key should be a guid, and I don't understand why you're having problems with your IPC calls finding your iModel. Can you show your code?
